libsolid-rs/src/auth.rs

```diff
@@ -382,20 +368,312 @@ pub mod webid_oidc {
+                    eprintln!(
+                        "[CallbackServer::index] processing incoming request on {}: {:#?}",
+                        &req.headers()
+                            .get("host")
+                            .ok_or_else(|| failure::err_msg(
+                                "could not find host in request headers"
+                            ))?
+                            .to_str()?,
+                        &req,
+                    );
+
+                    let query_string = req.query_string().to_string();
+
+                    let callback_values = actix_web::web::Query::<
+                        std::collections::HashMap<String, String>,
+                    >::from_query(&query_string)
+                    .map_err(|e| failure::err_msg(e.to_string()))?
+                    .into_inner();
+
+                    state
+                        .lock()
+                        .map_err(|e| failure::err_msg(format!("could not acquire lock: {}", e)))?
+                        .send(callback_values)?;
+
+                    Ok("Ok"
+                        .to_string()
+                        .with_status(actix_web::http::StatusCode::OK))
+                };
+
+                match responder() {
+                    Ok(ok) => ok,
+                    Err(e) => e
+                        .to_string()
+                        .with_status(actix_web::http::StatusCode::BAD_REQUEST),
+                }
+            }
+        }
+
+        pub(crate) type CallbackValues = HashMap<String, String>;
+
+        pub(crate) struct LoginChannels {
+            pub(crate) authorize_url_sender: Sender<url::Url>,
+            pub(crate) authorize_url_receiver: Receiver<url::Url>,
+
+            pub(crate) callback_values_sender: Sender<CallbackValues>,
+            pub(crate) callback_values_receiver: Receiver<CallbackValues>,
+        }
+
+        impl Default for LoginChannels {
+            fn default() -> Self {
+                use std::sync::mpsc::channel;
+                let callback_url_channel = channel::<url::Url>();
+                let callback_values_channel = channel::<CallbackValues>();
+
+                LoginChannels {
+                    authorize_url_sender: callback_url_channel.0,
+                    authorize_url_receiver: callback_url_channel.1,
+
+                    callback_values_sender: callback_values_channel.0,
+                    callback_values_receiver: callback_values_channel.1,
+                }
+            }
+        }
+    }
+
     #[cfg(test)]
     mod tests {
         use super::*;
 
         #[test]
         #[cfg(feature = "test-net")]
-        #[cfg(feature = "test-interactive")]
-        fn interactive_login_to_solid_community() -> Result<()> {
-            let callback_bind_addr = String::from("127.0.0.1:36666");
-
+        fn login_to_solid_community() -> Result<()> {
             let provider_url = "https://solid.community".to_string();
```

Comment:
Should we use an env var for this as well? That way the test environment has free choice which server to test against.

Reply:
Yes, definitely! It'd be nice if the test can pickup the `provider_url` from the environment. This way we can easily also test against `localhost` or other providers.

Reply:
it does now

Reply:
AWESOME! 🎉